src/gc.c

```diff
@@ -3084,7 +3083,7 @@ JL_DLLEXPORT void jl_gc_collect(jl_gc_collection_t collection)
     }
     gc_debug_print();
 
-    int8_t old_state = jl_gc_state(ptls);
+    int8_t old_state = ptls->gc_state;
     ptls->gc_state = JL_GC_STATE_WAITING;
```

Comment:
Why don't these need to be atomic?

Reply:
```suggestion
    jl_atomic_store_release(&ptls->gc_state, JL_GC_STATE_WAITING);
```

The read doesn't need to be atomic because it's only ever written from the thread itself. I agree this appears to need a release store. I'm slightly surprised TSAN didn't notice that too, though perhaps the race window was just very small, since `jl_safepoint_start_gc` is an acquire-release fence.